File: blind75/python/maximumproductsubarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0

        max_product = nums[0]
        min_product = nums[0]
        result = nums[0]

        for num in nums [1:]:
            temp_max = max_product
            max_product = max(num, num * max_product, num * min_product)
            min_product = min(num, num * temp_max, num * min_product)
            result = max(result, max_product)

        return result

File: blind75/c++/Maximumproductsubarray.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.empty()) return 0;

        int max_product = nums[0];
        int min_product = nums[0];
        int result = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            int temp_max = max_product;
            max_product = max(nums[i], max(nums[i] * max_product, nums[i] * min_product));
            min_product = min(nums[i], min(nums[i] * temp_max, nums[i] * min_product));
            result = max(result, max_product);
        }
        return result;
    }
};
